### PR TITLE
Add pre-commit hooks for bandit, black, codespell, and flake8

### PR DESCRIPTION
**Pull request by cclauss**
As discussed in #2268 it might be helpful to move four fast lint tests into a [.pre-commit-config.yaml](https://pre-commit.com) to halt git commits that contain errors.  Given that some contributors will not have pre-commit installed, this PR also includes a GitHub Action so that contributors and maintainers can see which tests have failed.

### CODE REVIEW

**Comment by mzjp2** [comment]
I would recommend pulling together the `pre-commit` job and the `lint_python` job into one, see [quantumblacklabs/kedro](https://github.com/quantumblacklabs/kedro) as an example, and the CI running [here](https://app.circleci.com/pipelines/github/quantumblacklabs/kedro/2767/workflows/08a54edf-7fbe-4303-b656-e58a2b8a3b18/jobs/48766)

**Comment by cclauss** [comment]
@mzjp2 Your review again please.

File: .github/workflows/lint_python.yml [hunk 1]
```diff
@@ -1,17 +1,25 @@
...
       - uses: actions/checkout@v2
       - uses: actions/setup-python@v2
-      - run: pip install bandit black codespell flake8 isort mypy pytest pyupgrade safety
-      - run: bandit --recursive --skip B101,B108,B301,B403,B404,B603 .
-      - run: black --check .
-      - run: codespell --ignore-words-list=nd,reacher,thist,ths -w
-      - run: flake8 --ignore=E203,E402,E712,E722,E731,E741,F401,F403,F405,F524,F841,W503
-                    --max-complexity=30 --max-line-length=456 --show-source --statistics
+      - run: pip install isort mypy pytest pyupgrade safety
       - run: isort --check-only --profile black . || true
```
Comment (mzjp2) [inline .github/workflows/lint_python.yml:23]: Do you wanna move `isort` into `pre-commit` as well (it works super nicely for sorting your imports when you run `pre-commit` locally)?

File: .github/workflows/lint_python.yml [hunk 2]
```diff
@@ -1,17 +1,25 @@
...
     steps:
       - uses: actions/checkout@v2
       - uses: actions/setup-python@v2
-      - run: pip install bandit black codespell flake8 isort mypy pytest pyupgrade safety
-      - run: bandit --recursive --skip B101,B108,B301,B403,B404,B603 .
-      - run: black --check .
-      - run: codespell --ignore-words-list=nd,reacher,thist,ths -w
-      - run: flake8 --ignore=E203,E402,E712,E722,E731,E741,F401,F403,F405,F524,F841,W503
-                    --max-complexity=30 --max-line-length=456 --show-source --statistics
+      - run: pip install isort mypy pytest pyupgrade safety
```
Comment (mzjp2) [inline .github/workflows/lint_python.yml:22]: also, maybe time to remove `pytest` from `lint_python` given that we now have a dedicated testing job?

File: .github/workflows/lint_python.yml (hunk 1 above)
Comment (cclauss) [reply .github/workflows/lint_python.yml:23]: pytest fails after running isort.

File: .github/workflows/lint_python.yml (hunk 2 above)
Comment (cclauss) [reply .github/workflows/lint_python.yml:22]: Given that the README installation instructions do not mention Docker, I was looking to install and run the tests without Docker.

**Comment by cclauss** [comment]
Removed isort, mypy, pytest, puupgrade, and safety

**Comment by jkterry1** [comment]
Are we sure that we want to require this to commit? the normal way of handling this is in tests, with a step to make sure people make sure their code is properly linted and so on. If we do this as a get check it will require users to have a bunch of software configured on their computers, which is surely going to create incident.

**Comment by cclauss** [comment]
The only thing that a contractor really should install is black.  The rest is will be caught by test and is optional for local installation.

**Comment by vwxyzjn** [comment]
Hey @cclauss thanks for making this contribution. Do you have a video demoing the workflow with `pre-commit`? I guess seeing that will help people better understand the PR.

**Comment by jkterry1** [comment]
Could you please also resolve the merge conflicts?

**Pull request merged**